Add script to plot transects from MALI output

trhille opened the pull request:
This merge adds a script to plot ice surfaces, bed topography, speed and optionally temperature along a transect. The transect can either be defined using `-x` and `-y` options from the command line, or using `-c` to point to a csv file containing x and y coordinates in the first and second columns, respectively, with no header row. Use `--temperature` if you wish to plot `temperature` along the transect. Note that this uses the 3D `temperature` field and ignores `surfaceAirTemperature` and `basalTemperature`, but these could be added in the future. You can plot more than one time index, specified using the `-t` option; however, plotting temperature at multiple times is not supported. The script increases the number of specified sample points to match the highest resolution in the mesh, based on `dcEdge`, and uses linear interpolation to produce high-resolution transects from a minimal number of specified points (minimum is 2). Temperature is interpolated in x and y, but not in the vertical dimension.

trhille commented:
Testing
-------
A transect down the middle of Thwaites Glacier from an ISMIP6 run in 2175:
`python plot_transect.py -d /pscratch/sd/t/trhille/ISMIP6/archive/expAE03/rst.2175-01-01.nc -x -1589311.171,-1208886.353 -y -473916.7182,-357411.6176  --temperature
<img width="642" alt="image" src="https://github.com/MPAS-Dev/MPAS-Tools/assets/17446278/c80b1e38-d1fc-4fd8-956d-09c327d89f36">

Transects down the middle of Thwaites Glacier from an ISMIP6 run in 2175 and 2179:
`python plot_transect.py -d /pscratch/sd/t/trhille/ISMIP6/archive/expAE03/output/output_state_2175.nc -t 0,-1 -x -1589311.171,-1208886.353 -y -473916.7182,-357411.6176 `
<img width="643" alt="image" src="https://github.com/MPAS-Dev/MPAS-Tools/assets/17446278/c4f56c8d-d3b1-45cc-bd6a-6d3296303c12">

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,214 @@
+import numpy as np
+import csv
+from netCDF4 import Dataset
+from optparse import OptionParser
```
Given that optparse is deprecated, we should probably use argparse for new scripts.

trhille replied:
`argparse` does not seem to be able to handle negative numbers in the `-x` and `-y` arguments. This seems to be a known bug that was possibly fixed a while ago, but `argparse.__version__` returns `1.1` for both `dev_compass_1.2.0-alpha.6` and `e3sm_unified_1.9.1_login`, while the latest version of argparse is 3.12.

matthewhoffman replied:
I agree that is a big enough annoyance to stick with optparse until we're forced to update.

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,214 @@
+
+import numpy as np
+import csv
+from netCDF4 import Dataset
```
Ideally we would use xarray to allow support for times from different files in a multi-file dataset.  But I'm happy to save that work for the first person who needs the functionality rather than making this review process take a bunch of time to make a perfect script.

trhille replied:
Yeah, let's leave that for future development because I don't have time to make that change now.

matthewhoffman replied:
Sounds good

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,214 @@
+if "surfaceSpeed" in dataset.variables.keys():
+    speed = dataset.variables["surfaceSpeed"][:] * 3600. * 24. * 365.
+else:
+    speed = np.sqrt(dataset.variables["uReconstructX"][:,:,0]**2. + 
```
Do you want to check if uReconstructX/Y are available?  It seems the script won't work if there is no velocity data in the output file.  I'd be fine committing it with the requirement that some form of velocity data are present (which is usually going to be the case), and saving changing that for if such a use case comes up.

trhille replied:
Resolved in 099bf72e

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,214 @@
+    speed *= 3600. * 24. * 365.  # convert from m/s to m/yr
+
+if options.interp_temp:
+    temperature = dataset.variables['temperature'][:]
```
Are you sure you need to read in the entire temperature array?  (Same goes for speed and thickness arrays above.)  For small files this is fine, but it might get inefficient for large files.  From what I'm seeing so far below, I think you can get away with dropping the `[:]` here and only accessing the data for the requested time levels when needed below.

trhille replied:
Addressed in commit 1d0e288.

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,214 @@
+    x = [float(i) for i in options.x_coords.split(',')]
+    y = [float(i) for i in options.y_coords.split(',')]
+
+# increase sampling by a factor of 100
```
Do you want to make this a function of dcEdge instead of always using 100?  (Another thing that could be saved for future improvement)

trhille replied:
Added in commit 48afa27.

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,214 @@
+if options.interp_temp:
+    temperature = dataset.variables['temperature'][:]
+    
+bedTopo = dataset.variables["bedTopography"][0,:]
```
Add a note it's using the initial bed only.  We may be having changing bed topography routinely soon, and when that's so, this script will be useful for looking at that.  For now, simply printing a warning about this would be fine with me.

matthewhoffman reviewed (changes requested):
@trhille , this looks like a great tool to have.  I would have suggested seeing what's being  done for transects in MPAS-Analysis to avoid reinventing the wheel, but there is enough ice-sheet-specific stuff in here (and you already have this done) that I'm not going to suggest that now.  I made a number of comments and questions, but I don't want to make this a really onerous review process.

trhille commented:
Thanks for the review @matthewhoffman! I'll get on these changes.

>  I would have suggested seeing what's being done for transects in MPAS-Analysis to avoid reinventing the wheel.

I already had most of this lying around in a branch from a few years ago, so really the new thing was the temperature plotting.

trhille commented:
@matthewhoffman, I've addressed all your comments in one form or another. The change to `argparse` rather than `optparse` is clunky enough that I'd rather leave it using `optparse`, even though it's deprecated.

matthewhoffman commented on landice/output_processing_li/plot_transect.py:
```diff
@@ -0,0 +1,220 @@
+    y = [float(i) for i in options.y_coords.split(',')]
+
+# increase sampling to match highest mesh resolution
+total_distance = np.sqrt( (x[0] - x[1])**2. + (y[0] - y[1])**2.)
```
Why not use the `distance` variable you calculate on line 132?  In addition to the fact that you are already calculating it, it would handle curved transects better.  And isn't this approach just considering the distance between the first two points, but it's possible to have many points defining a transect?

trhille replied:
Fixed in e303679.
